keras/src/layers/preprocessing/spectrogram.py

```diff
@@ -0,0 +1,325 @@
+                # easier to be compatible with multiple backends,
+                # and this is a one-time operation to store the constant
+                # window tensor in the graph
+                win = scipy.signal.get_window(
+                    win, self.frame_length, self.periodic
+                )
+            win = ops.convert_to_tensor(win, dtype=dtype)
+            if len(win.shape) != 1 or win.shape[-1] != self.frame_length:
+                raise ValueError(
+                    "The shape of `window` must be equal to [frame_length]."
+                    f"Received: window shape={win.shape}"
+                )
+            win = ops.reshape(win, [-1, 1, 1])
+            if self.scaling == "density":
+                scaling = ops.sqrt(ops.sum(ops.square(win)) + backend.epsilon())
+            elif self.scaling == "spectrum":
+                scaling = ops.abs(ops.sum(win)) + backend.epsilon()
+
+        if self.padding != "valid":
+            assert self.frame_length > 0
+            pad_value = self._padding_length // 2
+            inputs = ops.pad(
+                inputs,
+                [[0, 0], [pad_value, pad_value], [0, 0]],
+                mode=self.padding_mode,
+            )
+
+        real_signal = None
+        imag_signal = None
+        power = None
+        if self.mode != "imag":
+            real_kernel = ops.cast(
+                self.real_kernel if win is None else self.real_kernel * win,
+                dtype,
+            )
+            real_signal = (
+                ops.conv(
+                    inputs,
+                    real_kernel,
+                    strides=self._frame_step,
+                    data_format="channels_last",
+                )
+                / scaling
+            )
+
+        if self.mode != "real":
+            imag_kernel = ops.cast(
+                self.imag_kernel if win is None else self.imag_kernel * win,
+                dtype,
+            )
+            imag_signal = (
+                ops.conv(
+                    inputs,
+                    imag_kernel,
+                    strides=self._frame_step,
+                    data_format="channels_last",
+                )
+                / scaling
+            )
+        if self.mode == "real":
+            return real_signal
+        elif self.mode == "imag":
+            return imag_signal
+        elif self.mode == "angle":
+            return ops.arctan2(imag_signal, real_signal)
+        else:
+            power = ops.square(real_signal) + ops.square(imag_signal)
+
+        if self.mode == "psd":
+            return power + ops.pad(power[..., 1:-1], [[0, 0], [0, 0], [1, 1]])
+
+        linear_stft = ops.sqrt(power + backend.epsilon())
+
+        if self.mode == "magnitude":
+            return linear_stft
+        elif self.mode == "log":
+            return ops.log(backend.epsilon() + linear_stft)
+
+        raise NotImplementedError(f"{self.mode} mode is not implemented")
```

Comment:
Input validation should happen in the constructor not at call time

Reply:
I removed the validation in `__call__`, which was redundant since an equivalent check is already made in the constructor.